exec/Chainweb/BackfillTransfers.hs

```diff
@@ -0,0 +1,138 @@
+      unless (checkMinHeights minHeights) $ do
+        logg Error "Make sure transfers table has an entry for every chain id!"
+        exitFailure
+      let maxMinHeights = maximum $ mapMaybe snd $ minHeights
+      -- get maximum possible number of entries to fill
+      effectiveTotal <- withDbDebug env Debug $ runSelectReturningOne $ select $ bigEventCount maxMinHeights eventsActivationHeight
+      unless (isJust effectiveTotal) $ die "Cannot get the number of entries needed to fill transfers table"
+      mapM_ (\(cid, h) -> logg Info $ fromString $ printf "Filling transfers table on chain %d from height %d to height %d." cid eventsActivationHeight (fromJust h)) minHeights
+      ref <- newIORef 0
+      catch
+        (race_ (progress logg ref (fromIntegral $ fromJust $ effectiveTotal))
+            (forM_ (rangeToDescGroupsOf chunkSize (fromIntegral eventsActivationHeight) (fromIntegral maxMinHeights))
+             $ \(Low endingHeight, High startingHeight) ->
+                transferInserter ref (fromIntegral startingHeight) (fromIntegral endingHeight)))
+        (\(e :: SomeException) -> do
+            printf "\nDepending on the error you may need to run backfill for events\n%s\n" (show e)
+            exitFailure)
+  where
+    logg = _env_logger env
+    pool = _env_dbConnPool env
+    chunkSize = fromMaybe 200 $ _backfillArgs_chunkSize args
+    getValidTransfer :: Event -> (Sum Int, [Transfer] -> [Transfer])
+    getValidTransfer ev = maybe mempty ((Sum 1, ) . (:)) $ createTransfer ev
+    transferInserter :: IORef Int -> Int64 -> Int64 -> IO ()
+    transferInserter count startingHeight endingHeight = do
+        P.withResource pool $ \c -> withTransaction c $ runBeamPostgres c $ do
+          evs <- runSelectReturningList $ select $ eventSelector startingHeight endingHeight
+          let (Sum !cnt, tfs) = foldMap getValidTransfer evs
+          runInsert $
+            insert (_cddb_transfers database) (insertValues (tfs []))
+            $ onConflict (conflictingFields primaryKey) onConflictDoNothing
+          liftIO $ atomicModifyIORef' count (\cnt' -> (cnt' + cnt, ()))
+
+chainMinHeights :: Pg [(Int64, Maybe Int64)]
+chainMinHeights = runSelectReturningList $ select $ aggregate_ (\t -> (group_ (_tr_chainid t), min_ (_tr_height t))) (all_ (_cddb_transfers database))
+
+createTransfer :: Event -> Maybe Transfer
+createTransfer ev =
+      Transfer
+      <$> pure (_ev_block ev)
+      <*> pure (_ev_requestkey ev)
+      <*> pure (_ev_chainid ev)
+      <*> pure (_ev_height ev)
+      <*> pure (_ev_idx ev)
+      <*> pure (_ev_module ev)
+      <*> pure (_ev_moduleHash ev)
+      <*> from_acct
+      <*> to_acct
+      <*> amount
+  where
+    amount = case _ev_params ev ^? to unwrap . ix 2 . key "decimal" of
+      Just (A.Number n) -> Just $ toRealFloat n
+      _ -> case _ev_params ev ^? to unwrap . ix 2 . key "int" of
+        Just (A.Number n) -> Just $ toRealFloat n
```

Comment:
Apparently Postgres won't let you cast from jsonb to decimal (see https://www.postgresql.org/message-id/0154d35a-24ae-f063-5273-9ffcdf1c7f2e%40postgrespro.ru).  I'd like to verify that the SQL expressions we're using to convert data in the events table into the transfers table is handling this correctly.  Can we do a search for transactions that have an amount in the format `{"decimal": "1.234"}` and verify that those rows are indeed showing up in the transfers table?